src/common_union.rs

```diff
@@ -199,3 +200,109 @@ impl From<JsonUnionField> for ScalarValue {
         }
     }
 }
+
+pub struct JsonUnionEncoder {
+    boolean: BooleanArray,
+    int: Int64Array,
+    float: Float64Array,
+    string: StringArray,
+    array: StringArray,
+    object: StringArray,
+    type_ids: ScalarBuffer<i8>,
+}
+
+impl JsonUnionEncoder {
+    #[must_use]
+    pub fn from_union(union: UnionArray) -> Option<Self> {
+        if is_json_union(union.data_type()) {
+            let (_, type_ids, _, c) = union.into_parts();
+            Some(Self {
+                boolean: c[1].as_any().downcast_ref::<BooleanArray>().cloned()?,
+                int: c[2].as_any().downcast_ref::<Int64Array>().cloned()?,
+                float: c[3].as_any().downcast_ref::<Float64Array>().cloned()?,
+                string: c[4].as_any().downcast_ref::<StringArray>().cloned()?,
+                array: c[5].as_any().downcast_ref::<StringArray>().cloned()?,
+                object: c[6].as_any().downcast_ref::<StringArray>().cloned()?,
```

Comment:
It seems like allowing the downcasts to fail here is unnecessary due to the `is_json_union` check.

```suggestion
                boolean: c[1].as_boolean().clone(),
                int: c[2].as_primitive().clone(),
                float: c[3].as_primitive().clone(),
                string: c[4].as_string().clone(),
                array: c[5].as_string().clone(),
                object: c[6].as_string().clone(),
```